{%extends "base.html" %}
{% block title %} Documentation {% endblock %}
{% block content %}



<center>
<h2> Documentation </h2>
</center>


<div class="container text-right" style="padding-top: 50px;"> 
 <a href="{{url_for('doc_download')}}" class="btn btn-danger"> Download! </a>
</div> 


<div class="container" style="margin-top: 50px;">

    <center> <h3>Methodology</h3> </center>
    <br>
    <h3> Aim </h3>

        <br>

    <div class="para">


        <p>The SNPip webserver provides a broad database containing SNP information from chromosome 22. The main functions of SNPip relies on basic data search by chromosomal location, gene name and SNP name for a subset of 5 populations. The search query outputs basic statistics for allele frequency and genotype frequency for every subpopulation. Additionally, if multiple SNPs are selected it contains summary statistics for genetic diversity, haplotype diversity and Tajimas D. If multiple populations are selected it contains statistics of pairwise genetic distance and delta Tajimas-D.</p> 

        <br>

        <p>Search by:</p> 

        <br>

        <ul>
            <li> Gene annotation </li>
            <li> SNP annotation </li>
            <li>Reference and Alternate allele </li>
        </ul>

        <img src="/static/search.gif">



        <br>

        <p>Output:</p> 
        <br>
        <ul>
            <li> Alternate allele frequency  </li>
            <li> Genotype frequency </li>
            <li> Shannon Diversity  </li>
            <li> Haplotype Diversity </li>
            <li> Tajimas-D  </li>
            <li> Delta Tajimas-D between two populations </li>
            
        </ul>


        <br>

        <h3> Data </h3>

        <br>

        <p>
            SNPip uses 1000 Genomes Project Phase 3 variants and also its definition of the super populations of European, East Asian, Southeast  Asian, South American and African. Super populations are defined using the 1000G panel file. <br>

            The database contains biallelic SNPs derived from pre-processing the phase 3 genotype data. SNPip contains only single nucleotide variants (SNPs). Indels and larger structural variants have been removed. <br>

            The data only contains high confidence calls for ALT alleles and positions that have passed all filters.<br>

            <span class="text-danger"> <b> SNPip contains SNPs located on chromosomes 22 (February 2022) </b> </span>
        </p>



        <p>
            All the 1000G variants has been preprocessed with bcftools before creating the database using the following criteria: 
        </p>
        <br>

        <ol>
            <li>Remove SNPs with duplicate rsIDs. </li>
            <li>Remove Indels, and multiallelic sites </li>
            <li> Run PLINK variant filtering </li>
            <li> Remove SNPs with duplicate chromosomal coordinates.   </li>
        
        </ol>

        <br>

        <h4> Population and Sub-population </h4>

        <p> SNPip super population definitions are based on the following 1000G cohorts: </p>


        <ul>
            <li> European (EUR) </li>
                <ul>
                    <li style="color: green;"> <b> British in England and Scotland (GBR) </b> </li>
                </ul>
            <li> East Asian (EAS) </li>
                <ul>
                    <li style="color: green;"> <b>Han Chinese in Beijing, China (CHB) </b> </li>
                </ul>
            <li> Africa (AFR) </li>
                <ul>
                    <li style="color: green;"> <b>Esan in Nigeria (ESN) </b> </li>
                </ul>           
            <li> Southeast  Asian (SAS)  </li>
                <ul>
                    <li style="color: green;"> <b>Bengali in Bangladesh (BEB) </b> </li>
                </ul>
            <li> American Ancestry (AMR)  </li>
                <ul>
                    <li style="color: green;"> <b>Peruvian in Lima, Peru (PEL) </b> </li>
                </ul>
        </ul>

        <br>

        <h3> Statistics </h3>

        <br>

        <h4> Shannon’s Diversity index </h4>


        <p>
            Shannon diversity index (SDI) is widely used in ecology to estimate species diversity and can be applied to population genetic studies. SNpip calculates genetic diversity using SDI; the allele frequency for a SNP variant in each population is used to estimate the genetic diversity between each population. 
        </p>



        <p style="color: green;"> H = -∑[(pi) * ln(pi)] </p>

        <ul>
            <li>H = Shannon Diversity index </li>
            <li>pi = allele frequency(af) where the proportion of individuals with a particular alternate allele(n) found divided by the total number of individual alleles(N) found.</li>
            <ul>
                <li>pi = n/N    or  af = al/ an </li>    
                <li>al = allele count   an = allele number </li>
            </ul>
            <li>∑= sum of the calculation </li> 
            <li>ln = natural logarithm </li>
        </ul> 


        <p> SDI calculated using a python algorithm with the SDI formula:   </p>

        <ol>
            <li>Select allele frequency for the SNP variant for each population.</li>
            <li>For each population, multiply the proportion by the logarithm of the proportion.</li>
            <li>Sum all the numbers from step 2.</li>
            <li>Multiply the sum by -1.</li>
        </ol>

        <br>

        <h3> Output files </h3>

        <br>

        <ul>
            <li> Hudson_FST.csv   </li>
                <ul>
                    <li> If 2 or more populations are selected during search there will be statistics otherwise will be an empty file. </li>
                </ul>
            <li> Tajima_D.csv </li>
            <li> Shannon_Diversity.csv </li>
            <li> Haplotype_Diversity.csv </li>
        </ul>

  
  
  


    </div>






</div>









{% endblock %}